bigquery/api/installed_app.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+
+# Copyright 2015, Google, Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Command-line application that demonstrates using BigQuery with credentials
+obtained from an installed app.
+
+This sample is used on this page:
+
+    https://cloud.google.com/bigquery/authentication
+
+For more information, see the README.md under /bigquery.
+"""
+# [START all]
+
+import argparse
+import pprint
+
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from oauth2client import tools
+from oauth2client.client import AccessTokenRefreshError
+from oauth2client.client import flow_from_clientsecrets
+from oauth2client.file import Storage
+
+SCOPES = ['https://www.googleapis.com/auth/bigquery']
+
+
+def main(args):
+    storage = Storage('credentials.dat')
+    credentials = storage.get()
+
+    if credentials is None or credentials.invalid:
+        flow = flow_from_clientsecrets(
+            args.client_secrets, scope=SCOPES)
+        # run_flow will prompt the user to authorize the application's
+        # access to BigQuery and return the credentials.
+        credentials = tools.run_flow(flow, storage, args)
+
+    # Create a BigQuery client using the credentials.
+    bigquery_service = discovery.build(
+        'bigquery', 'v2', credentials=credentials)
+
+    # List all datasets in BigQuery
+    try:
+        datasets = bigquery_service.datasets()
+        listReply = datasets.list(projectId=args.project_id).execute()
+        print('Dataset list:')
+        pprint.pprint(listReply)
+
+    except HttpError as err:
+        print('Error in listDatasets:')
+        pprint.pprint(err.content)
+
+    except AccessTokenRefreshError:
+        print('Credentials have been revoked or expired, please re-run'
+              'the application to re-authorize')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        # Use oauth2client's argparse as a base, so that the flags needed
+        # for run_flow are available.
+        parents=[tools.argparser])
+    parser.add_argument(
+        'project_id', help='Your Google Cloud Project ID.')
+    parser.add_argument(
+        'client_secrets', help='Path to your project\'s client_secrets.json')
```

Comment:
> this requires the developer who's reading this to have to place a file in a particular location or change a global variable in the code to run.

Agreed - it's not great. I think it's appropriate for an installed app, though. I'm trying to think of scenarios where you'd even be using bigquery as an installed app, with 3-legged oauth. The only scenario I can think of is if it's an internal tool you distribute to folks who are already project members, to access internal bq tables. In that case, I would think you'd want to embed your project information into the script before distributing it, rather than tell people to pass the script flags all the time.

Just thought of another situation - what if the bq tables were ACL'ed to allow a given user, but they're not part of the project at all? In that case, you'd still want 3LO for access control, but the request is still tied to the developer's project. In that case, the user shouldn't know or care about the project at all.

Re: client secret / project id

Our [oauth2 docs](https://developers.google.com/identity/protocols/OAuth2InstalledApp) seem to say the client secret and project id are tied together. AFAICT, all the user does is log in (and presumably some project admin adds that user to the project, or an ACL or something). Can you point me to docs that say otherwise?